module/zfs/dmu_send.c

```diff
@@ -1995,7 +1996,24 @@ create_begin_record(struct dmu_send_params *dspp, objset_t *os,
 	drr->drr_u.drr_begin.drr_flags |= DRR_FLAG_SPILL_BLOCK;
 
 	dsl_dataset_name(to_ds, drrb->drr_toname);
-	if (!to_ds->ds_is_snapshot) {
+	if (dspp->partialok) {
+		objset_t *mos = dmu_objset_pool(os)->dp_meta_objset;
+
+		/*
+		 * If we're sending a partially received dataset
+		 * we need to get the toguid and toname from the
+		 * receive resume token since to_ds is a
+		 * non-snapshotted, inconsistent dataset.
+		 */
+		VERIFY0(zap_lookup(mos, dmu_objset_id(os),
+		    DS_FIELD_RESUME_TOGUID, sizeof (drrb->drr_toguid), 1,
```

Comment:
This code asserts that these fields exist (i.e. that this is a partially received snapshot/filesystem).  Where does the kernel check that?

Reply:
this file is part of the kernel module

Reply:
Do you think we should zap_lookup TOGUID and TONAME (if partialok) in dmu_send as part of preflight checking?  I think Tom didn't want to introduce an error handling path here in create_begin_record hence the VERIFYs.  But we could do it in dmu_send before we have too much to unwind.  @mahrens

Reply:
Yeah I think we need to do the checking somewhere in the kernel. `dmu_send()` seems fine to me.

Reply:
We check it in dmu_send_impl, line 2336.